### ISSUE TITLE
Posterior expectation of the ambient profile in EmptyDrops

### ISSUE THREAD

**Issue by paula-tataru** [MarioniLab/DropletUtils#39]
Hi,

I am looking into the details of EmptyDrops and I have a bit of a hard time understanding a detail regarding the approach used.

In the article, you state that using the Good-Turing algorithm "ensures that genes with zero counts in the ambient pool have non-zero proportions, avoiding undefined likelihoods in downstream calculations. " and that the likelihood equation "highlights the usefulness of pg instead of a naive estimate of the proportions from A, as the former is guaranteed to be non-zero such that Lb is defined."

However, looking into the Good-Turing algorithm, it is clear that the estimated posterior for a gene with ambient expression 0 is non-zero if and only if there is at least one gene with ambient expression 1. I have verified this by using goodTuringProportions from edgeR.

So the first question is, how likely is it for that to happen and end up with undefined likelihood?

The second question is, are you dealing with this situation separately in the code? I have generated data that should lead to undefined likelihoods, however the emptyDrops function seems to work just fine in this situation. Here is the code I used.

```R
noGenes <- 100
noBarcodes <- 1000
expressionMatrix <- sapply(1:noBarcodes, function(i) {rnbinom(noGenes, mu = 100, size = 1/50)})

totalExpression <- colSums(expressionMatrix)
summary(totalExpression)

lower <- round(quantile(totalExpression, probs = 0.25))
ambientBarcodes <- which(totalExpression <= lower)
remainingBarcodes <- setdiff(1:noBarcodes, ambientBarcodes)
length(ambientBarcodes)

# Add genes with expression 0 in the ambient profile
noSpecialGenes <- 10
specialGenes <- matrix(0, ncol = noBarcodes, nrow = noSpecialGenes)
specialGenes[, remainingBarcodes] <- sapply(1:length(remainingBarcodes), function(i) {rnbinom(noSpecialGenes, mu = 100, size = 1/50)})
summary(rowSums(specialGenes))
expressionMatrix <- rbind(expressionMatrix, specialGenes)

ambientProfile <- rowSums(expressionMatrix[, ambientBarcodes])
summary(ambientProfile)
posterior <- goodTuringProportions(ambientProfile)
summary(posterior)

# Remove any gene with expression 1 in the ambient profile
pos <- which(ambientProfile == 1)
summary(pos)
if (length(pos) > 0) {
    expressionMatrix <- expressionMatrix[-pos, ]
}

testing <- emptyDrops(expressionMatrix, lower = lower)
summary(testing$PValue)

```

Thank you!
Paula

**Comment by LTLA** [comment]
> So the first question is, how likely is it for that to happen and end up with undefined likelihood?

It's never happened to me, so pretty unlikely, I would say. Might explain things like #33, though.

> The second question is, are you dealing with this situation separately in the code?

No. It's not clear what the best course of action would be here. Perhaps we may approximate the Good-Turing estimate by pretending we have a single transcript with a count of 1; the resulting estimates will be extremely low (the correct state of affairs) but at least they will be non-zero.

> however the emptyDrops function seems to work just fine in this situation.

You'll notice the log-probabilities are all `-Inf`, though, which is closest to the "correct" output here.

**Comment by paula-tataru** [comment]
Hi Aaron,

Thanks a lot for your answer. I didn't look at the likelihoods, it's true, just at the p-values, but now I understand how they fit together :)

**Comment by LTLA** [comment]
Well, this should be fixed in 73a432c8ef83546b5b9dc15608f28a66a6b8959a. The hack is to add a "pseudo-feature" with a count of 1, manually calculate the Good-Turing estimates for zero-count features in the presence of the pseudo-feature, and then replace the zeros in the edgeR GT output with the manual estimates (with some scaling so that everything sums to 1).

The implementation I've chosen has at least some properties:

- The relative values of probabilities for non-zero counts are not changed from the edgeR GT output.
- The probabilities for zero-count features are the same as what you would get for those features if you "redistributed" counts across genes so you just happened to get a single gene with 1 count.

Not sure how nice these properties are, but for an _ad hoc_ method, it's something at least.
